Add support for reading from a raw fru dump

matsievskiysv opened the pull request:
This is a squashed and rebased PR #9.
It works for the most part, except for custom fields, time and <s>multi-records</s>.

Source file:
```json
{
	/* The internal use area, if specified, must list all the data as hex string,
	   the length is calculated automatically */
	"internal" : "010203040A0B0C0D",
	"chassis" : {
		"type": 10,
		"pn" : "CHAS-C00L-12",
		/* Fields may be given explicit types by setting to object with keys `type` and `data`*/
		/* Supported types are: "bcdplus", "6bitascii" and "text" */
		"serial": {
		    "type": "bcdplus",
		    "data": "45678"
		},
		"custom" : [
			{ "data" : "Auto-typed text custom field" },
			{ "type" : "binary", "data": "B14A87" },
			/* For explicit text types range is not checked, so be careful */
			{ "type" : "bcdplus", "data": "1234" },
			{ "type" : "6bitascii", "data": "1234" },
			{ "type" : "text", "data": "1234" }
		]
	},
	"board" : {
		/* The date, if not specified, will be taken automatically from
		 * the current system time. You may specify the `-u` option to
		 * `frugen` in order to leave the date 'Unspecified' */
		/* "date" : "1/10/2016 3:00:45",*/
		"mfg" : "Biggest International Corp.",
		"pname" : "Some Cool Product",
		"serial" : "123456",
		"pn" : "BRD-PN-345",
		"file" : "example1.json",
		"custom" : [
			{ "type" : "binary", "data" : "0123DEADBABE" },
			{ "type" : "auto",   "data" : "This is a text custom field" },
			{ "type" : "auto",   "data" : "This is test2" }
		]
	},
	"product" : {
		"lang": 1,
		"mfg" : "Super OEM Company",
		"pn" : "PRD-PN-1234",
		"pname" : "Label-engineered Super Product",
		"serial" : "OEM12345",
		"atag" : "Accounting Dept.",
		"ver" : "v1.1",
		"file" : "example2.json",
		"custom" : [
			{ "type" : "auto",   "data" : "Product Custom 1" },
			{ "type" : "auto",   "data" : "PRDCSTM" },
			{ "type" : "auto",   "data" : "PRDCSTM2" },
			{ "type" : "binary", "data" : "C001BEEF" }
		]
	},
	"multirecord" : [
		{ "type" : "management", "subtype" : "uuid", "uuid" : "9bd70799-ccf0-4915-a7f9-7ce7d64385cf" }
	]
}
```

Is converted using commands:
```bash
./build/frugen-static -j -z example.json example.bin
./build/frugen-static -r -z example.bin
```

Output:
```
Chassis
        type: 10
        pn(sixbitascii): CHAS-C00L-12
        serial(bcdplus): 45678
        custom: Auto-typed text custom field
        custom:  J
        custom: 4
        custom:  4Q
        custom: 4
Product
        lang: 25
        mfg(text): Super OEM Company
        pname(text): Label-engineered Super Product
        serial(sixbitascii): OEM12345
        pn(sixbitascii): PRD-PN-1234
        ver(text): v1.1
        atag(text): Accounting Dept.
        file(text): example2.json
        custom: Product Custom 1
        custom:  L 3
        custom:  L 3 J
        custom:
Board
        lang: 25
        time: 13/55/2067 08:07:24
        mfg(text): Biggest International Corp.
        pname(text): Some Cool Product
        serial(bcdplus): 123456
        pn(sixbitascii): BRD-PN-345
        file(text): example1.json
        custom: #ޭ
        custom: This is a text custom field
        custom: This is test2
Multirecord
```

matsievskiysv commented:
Fixed custom fields:
```
Chassis
	type: 10
	pn(sixbitascii): CHAS-C00L-12
	serial(bcdplus): 45678
	custom(text): Auto-typed text custom field
	custom(binary): b14a87
	custom(bcdplus): 1234
	custom(sixbitascii): 1234
	custom(text): 1234
Product
	lang: 25
	mfg(text): Super OEM Company
	pname(text): Label-engineered Super Product
	serial(sixbitascii): OEM12345
	pn(sixbitascii): PRD-PN-1234
	ver(text): v1.1
	atag(text): Accounting Dept.
	file(text): example2.json
	custom(text): Product Custom 1
	custom(sixbitascii): PRDCSTM
	custom(sixbitascii): PRDCSTM2
	custom(binary): c001beef
Board
	lang: 25
	time: 13/29/2067 10:07:24
	mfg(text): Biggest International Corp.
	pname(text): Some Cool Product
	serial(bcdplus): 123456
	pn(sixbitascii): BRD-PN-345
	file(text): example1.json
	custom(binary): 0123deadbabe
	custom(text): This is a text custom field
	custom(text): This is test2
Multirecord
```

matsievskiysv commented:
Fixed time decoding:
```
time: 15/11/2022 09:35:33
```

matsievskiysv commented:
I think, the only thing is left to do for now is JSON dump.

matsievskiysv commented:
I think, it's ready for review.

## Basic command usage
1. Generate FRU dump from `example.json`:
  ```bash
  fugen-static -j -z example.json example.bin
  ```
2. Parse FRU dump to [`jq`](https://stedolan.github.io/jq/) or `example2.json`
  ```bash
  frugen-static -r -z example.bin > example2.json
  frugen-static -r -z example.bin | jq
  ```
3. ...
4. Profit!

More involved example: parse FRU dump, replace one of the fields and save to JSON file
```bash
frugen-static -r -z example.bin | jq '.board.file.data |= "otherexample.json"' > otherexample.json
```

## Motivation

> The motivation for this is to use ipmitool to dump the current fru contents and then modify/add fields before writing it back. (From #9)

As may be seen from the examples above, dumping parsed data to stdout is more flexible than just using FRU dump in new dump generation:
* data can be examined
* data can be easily modified

## Further work

Internal area and multi-record parsers are not implemented yet.

matsievskiysv commented:
It's not very convenient to use functions defined in `fru_reader.h` as they work with file descriptor. It would be more flexible to expose functions that operate on a buffer (and maybe a set of helper functions for file descriptor and/or stream access).

matsievskiysv commented:
Refactored decoder functions a bit for better library API

AlexanderAmelkin reviewed (changes requested):
First of all, thank you for the effort! Then, there are some things I'd like to improve.

matsievskiysv commented on fru.c:
```diff
@@ -327,58 +404,29 @@ fru_field_t * fru_encode_data(int len, const uint8_t *data)
- *
- * For binary data use FRU_FIELDDATALEN(field->typelen) to find
- * out the size of the returned buffer.
- */
```
I've moved it into fru.h

matsievskiysv commented on frugen.c:
```diff
@@ -674,6 +748,55 @@ int main(int argc, char *argv[])
 #endif
 				}
+				else if (use_binary) {
+					int fd = open(optarg, O_RDONLY);
```
I think that these functions belong to fru library.

AlexanderAmelkin replied:
fair enough

matsievskiysv commented on frugen.c:
```diff
@@ -816,120 +939,245 @@ int main(int argc, char *argv[])
+		}
+		// dump to stdout
+		json_object_to_fd(fileno(stdout), json_root, 0);
+		json_object_put(json_root);
```
This approach is more limiting because it does not allow decoding binary into JSON. I, for one, need exactly this functionality.

AlexanderAmelkin replied:
Why? If you add a dump option, then it will dump to JSON the FRU constructed from any source, just like you need. If the dump option is implemented as specifying `-` for the output file, you won't even need a separate dry-run option. For your task you would just run `frugen -rz template.bin -`. Am I missing anything?

matsievskiysv replied:
So, just to clear, how exactly input options should be handled?

This way
|Command|Input|Output|
|:--|:--|:--|
|`./frugen --json --from-file template.json output.bin`|JSON|Bin|
|`./frugen --raw --from-file template.bin output.bin`|Bin|Bin|
|`./frugen --json --from-file template.json -`|JSON|Bin(stdout)|
|`./frugen --raw --from-file template.bin -`|Bin|Bin(stdout)|
|`./frugen --json --from-file template.json --dump output.json`|JSON|JSON|
|`./frugen --raw --from-file template.bin --dump output.json`|Bin|JSON|
|`./frugen --json --from-file template.json --dump -`|JSON|JSON(stdout)|
|`./frugen --raw --from-file template.bin --dump -`|Bin|JSON(stdout)|

Or this way
|Command|Input|Output|
|:--|:--|:--|
|`./frugen --json --from-file template.json output.bin`|JSON|Bin|
|`./frugen --raw --from-file template.bin output.bin`|Bin|Bin|
|`./frugen --json --from-file template.json -`|JSON|JSON(stdout)|
|`./frugen --raw --from-file template.bin -`|Bin|JSON(stdout)|

AlexanderAmelkin replied:
I think the latter is very concise and convenient and I'm fine with that approach. Do you have any objections?

matsievskiysv replied:
> Do you have any objections?

No, I just thought it would be better to agree on this before coding.

matsievskiysv commented:
I've pushed edits, resolving most of the notes. Later, I will squash commits and split changed into separate commits, but first, library API and CLI arguments need to be thought over.

AlexanderAmelkin commented on fru.c:
```diff
@@ -39,11 +39,39 @@
+    [FIELD_TYPE_AUTO] = "auto",
+    [FIELD_TYPE_BINARY] = "binary",
+    [FIELD_TYPE_BCDPLUS] = "bcdplus",
+    [FIELD_TYPE_SIXBITASCII] = "sixbitascii",
```
We still have `6bitascii` in `example.json` and in `frugen.c` at original line 278.
For compatibility reasons I would prefer to use that name instead of `sixbitascii`.

AlexanderAmelkin commented on fru.c:
```diff
@@ -590,6 +677,43 @@ fru_info_area_t *fru_create_info_area(fru_area_type_t atype,    ///< [in] Area t
+		size_t length = FRU_FIELDDATALEN(field->typelen);
+		custom_field->rec = calloc(1, FRU_FIELDMAXARRAY);
+		custom_field->rec->typelen = field->typelen;
+                if (FRU_ISTYPE(field->typelen, BINARY)) {
```
Instead of nesting `if`s, could you please use `switch(FRU_TYPE(field->typelen))` ?

AlexanderAmelkin commented on fru.c:
```diff
@@ -682,6 +831,64 @@ fru_board_area_t * fru_board_info(const fru_exploded_board_t *board) ///< [in] E
+		uint32_t val;
+		uint8_t arr[4];
+	} min_since_1996_big_endian;
+	min_since_1996_big_endian.val = 0;
```
could as well add ` = { 0 };` at the end of the declaration above.
I don't insist though.

AlexanderAmelkin replied:
you don't need to assign `0` to `val` after `min_since_1996_big_endian = {0}`, also please make it `{ 0 }`.

AlexanderAmelkin commented on frugen.c:
```diff
@@ -379,6 +388,40 @@ bool json_fill_fru_mr_reclist(json_object *jso, fru_mr_reclist_t **mr_reclist)
+			goto TYPE_STRING_ERR;
+		if ((entry = json_object_new_object()) == NULL)
+			goto ENTRY_ERR;
+		if (json_object_object_add(entry, "type", type_string))
```
This function is `void`, at least in my version of `json-c`.

AlexanderAmelkin commented on frugen.c:
```diff
@@ -816,120 +923,245 @@ int main(int argc, char *argv[])
+			json_object_object_add(json_root, "board", section);
+		}
+		// dump to stdout
+		json_object_to_fd(fileno(stdout), json_root, 0);
```
This function doesn't exist in my version of json-c. Probably, you need to demand some minimum version of the library in CMakeLists.txt? Better though, make the code compatible with version 0.11 for compatibility with legacy distributions like Ubuntu 16.04 (xenial).

Also, your patches produce a lot of new warnings for Ubuntu 16.04. That may be the same issue as above, but anyway please make sure that the code builds without warnings, or at least without new ones.

matsievskiysv replied:
I used json-c 0.16-2. I'll double-check the code with Ubuntu 16.04

matsievskiysv commented:
`frugen` is coupled too tight to the `fru` library, moving parsing functions from `frugen` to `fru` requires a lot of changes. This adds to an already big patch. @AlexanderAmelkin, maybe you could create a new branch for the reader feature, so these things could be added one PR at a time?

AlexanderAmelkin commented on fru.c:
```diff
@@ -272,11 +307,100 @@ static unsigned char *fru_decode_6bit(const fru_field_t *field)
+		  break;
+		case 0xF:
+		  out[i] = '?';
+		  break;
```
No need to duplicate the code, you could combine the cases:
```
case 0xD:
case 0xE:
case 0xF:
    out[i] = '?';
    break;
```

AlexanderAmelkin commented on fru.c:
```diff
@@ -590,6 +677,47 @@ fru_info_area_t *fru_create_info_area(fru_area_type_t atype,    ///< [in] Area t
+		field = (fru_field_t*)data;
+
+		// end of fields
+		if (field->typelen == 0xc1)
```
please create a macro of constant for `0xc1`, like `#define FRU_TYPE_EOF 0xc1`

AlexanderAmelkin commented on fru.c:
```diff
@@ -590,6 +677,47 @@ fru_info_area_t *fru_create_info_area(fru_area_type_t atype,    ///< [in] Area t
+		custom_field->rec = calloc(1, FRU_FIELDMAXARRAY);
+		custom_field->rec->typelen = field->typelen;
+		switch (FRU_TYPE(field->typelen)) {
+			case __TYPE_BINARY: {
```
I don't think those braces are needed here around case blocks, please remove them

AlexanderAmelkin commented on fru.c:
```diff
@@ -272,11 +307,100 @@ static unsigned char *fru_decode_6bit(const fru_field_t *field)
+		c = (field->data[i / 2] >> ((i % 2) ? 0 : 4)) & 0x0F;
+		switch (c) {
+		case 0xA:
+		  out[i] = ' ';
```
Why are the block contents using mixed indentation? Please use tabs only for indentation. Spaces can be used for alignment.

AlexanderAmelkin commented on fru.c:
```diff
@@ -272,11 +307,100 @@ static unsigned char *fru_decode_6bit(const fru_field_t *field)
+ */
+static
+bool fru_decode_bcdplus(const fru_field_t *field,
+			       uint8_t *out,
```
Please use spaces for alignment, don't use tabs.

AlexanderAmelkin commented on fru.c:
```diff
@@ -911,6 +1122,63 @@ fru_mr_area_t *fru_mr_area(fru_mr_reclist_t *reclist, size_t *total)
+
+	field = (fru_field_t*)data;
+	if (!fru_decode_data(field, &product_out->mfg,
+                         sizeof(product_out->mfg.val)))
```
What is this second line aligned to? I don't see it.
If this is not an alignment, then please use tabs.
When mixing indentation with alignment (like here), please use tabs for indentation and spaces for alignment. That way it wont visually break in editors that use a different tab size:

```c
	if (!fru_decode_data(field, &product_out->mfg,
	                     sizeof(product_out->mfg.val)))
```

This applies to the below cases as well.

AlexanderAmelkin replied:
Sorry, what you did is not exactly what I meant by 'tabs for indentation, spaces for alignment'.
If you change tab size, this is still going to break.
This is 1 level on indent, hence it must be just 1 tab. The rest must be spaces.

There are lots of other places like this too.

AlexanderAmelkin commented on fru.c:
```diff
@@ -1073,8 +1388,8 @@ void test_encodings(void)
-		out = fru_decode_data(field);
-		if (!out) {
+		if (!fru_decode_data(field->&typelen, &field->data,,
+                             FRU_FIELDMAXARRAY)) {
```
problems with indentation/alignment again

AlexanderAmelkin commented on frugen.c:
```diff
@@ -674,6 +761,55 @@ int main(int argc, char *argv[])
+					free(buffer);
+				}
 				else {
 					fatal("The requested input file format is not supported");
```
Now it looks more like `"Please specify the input file format"`

AlexanderAmelkin replied:
You didn't actually resolve this, did you?

matsievskiysv replied:
Sorry, this edit got lost during the commit split

AlexanderAmelkin commented on frugen.c:
```diff
@@ -816,120 +952,245 @@ int main(int argc, char *argv[])
+			}
+			json_object_object_add(section, "custom", temp_obj);
+			json_object_object_add(json_root, "board", section);
+		}
```
Looks like `if (has_multirec)` is missing from the JSON dump part and needs to be added (with proper decoding of everything that can be encoded, and a hex dump of what isn't yet encodeable). Probably as a separate PR in the branch that I've created for this feature set as you asked.

Please at least add some stub in this PR for that case.

Commit:
Add reader

Commit:
Reindent files

Commit:
Fix field comment line

Commit:
Fix wrong arguments to fatal macro

matsievskiysv commented:
I've updated the code according to your comments and split it into separate commits.
The following remains for the following PRs:
1. Extract JSON and raw reader logic into separate function. This requires a lot of changes because of statically allocated data in the `main` function. It would be easier to navigate in diff in separate PR.
2. Change CLI behaviour (from the table in previous messages). It makes sense to do this together with the previous task.

I propose to leave these improvements for the separate PR.

AlexanderAmelkin commented:
> I've updated the code according to your comments and split it into separate commits. The following remains for the following PRs:
> 
> 1. Extract JSON and raw reader logic into separate function. This requires a lot of changes because of statically allocated data in the `main` function. It would be easier to navigate in diff in separate PR.
> 2. Change CLI behaviour (from the table in previous messages). It makes sense to do this together with the previous task.
> 
> I propose to leave these improvements for the separate PR.

I agree that those changes can go in separate PRs.

matsievskiysv commented:
Since there're a lot of indentation problems, maybe it would be better to add a formatting rules for [clang-format](https://clang.llvm.org/docs/ClangFormat.html) or [indent](https://www.gnu.org/software/indent/manual/indent.html). 
This is a clang-format file I use for my projects and It's close to the format of the project (drop the `.txt` extension):
[.clang-format.txt](https://github.com/ipmitool/frugen/files/10408172/default.clang-format.txt)

AlexanderAmelkin commented:
> Since there're a lot of indentation problems, maybe it would be better to add a formatting rules for [clang-format](https://clang.llvm.org/docs/ClangFormat.html) or [indent](https://www.gnu.org/software/indent/manual/indent.html)

As far as I'm aware, clang-format doesn't support the concept of 'tabs for indent, spaces for alignment', and I wouldn't like to use any less. If you know how to teach it that, please tell.

I will look into `indent`'s capabilities.

matsievskiysv commented:
> clang-format doesn't support the concept of 'tabs for indent, spaces for alignment'

I think it is using it: [fru.c.txt](https://github.com/ipmitool/frugen/files/10435887/fru.c.txt)


![image](https://user-images.githubusercontent.com/26824900/212921599-c4948dfc-a533-4dcd-861a-2762c41587e9.png)
![image](https://user-images.githubusercontent.com/26824900/212921705-a9f6ef96-492e-4fb6-a875-d567106efa62.png)
